lib/request-proxy-send.js

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var numOrDefault = require('./util.js').numOrDefault;
+var rawHead = require('./request-proxy-util.js').rawHead;
+var strHead = require('./request-proxy-util.js').strHead;
+var TypedError = require('error/typed');
+
+var MaxRetriesExceeded = TypedError({
+    type: 'ringpop.request-proxy.max-retries-exceeded',
+    message: 'Max number of retries exceeded. ({maxRetries}) were attempted',
+    maxRetries: null
+});
+
+var RetryDivergenceError = TypedError({
+    type: 'ringpop.request-proxy.keys-diverged',
+    message: 'Destinations for proxied request have diverged. These keys: \n' +
+        ' {keys}, were originally intended for {origDestination}, but are\n' +
+        ' now meant for {newDestinations}.',
+    keys: null,
+    origDestination: null,
+    newDestinations: null
+});
+
+var RETRY_SCHEDULE = [0, 1, 3.5];
+
+function RequestProxySend(opts) {
+    this.ringpop = opts.ringpop;
+    this.requestProxy = opts.requestProxy;
+    this.keys = opts.keys;
+    this.channelOpts = opts.channelOpts;
+    this.request = opts.request;
+    this.retries = opts.retries;
+
+    this.retrySchedule = this.retries.schedule || RETRY_SCHEDULE;
+    this.maxRetries = numOrDefault(this.retries.max, this.retrySchedule.length);
+    this.maxRetryTimeout = this.retrySchedule[this.retrySchedule.length - 1] * 1000;
+
+    this.destinations = [this.channelOpts.host];
+    this.errors = [];
+    this.numRetries = 0;
+    this.reqStartTime = Date.now();
+    this.retryStartTime = null;
+    this.timeout = null;
+}
+
+RequestProxySend.prototype.abortRetry = function abortRetry(dests, callback) {
+    this.ringpop.stat('increment', 'requestProxy.retry.aborted');
+    this.ringpop.logger.warn('ringpop request proxy retry aborted after keys destinations diverged', {
```

Comment:
this log statement assumes implementation details that do not match the `abortRetry` method name.

Consider calling this `abortOnKeyDivergence` or changing the log statement